src/Domain/CoordinateMaps/CoordinateMap.hpp

```diff
@@ -98,10 +98,12 @@ class CoordinateMapBase : public PUP::able {
   virtual boost::optional<tnsr::I<double, Dim, SourceFrame>> inverse(
       tnsr::I<double, Dim, TargetFrame> target_point,
       double time = std::numeric_limits<double>::signaling_NaN(),
-      const std::unordered_map<std::string,
-                               domain::FunctionsOfTime::FunctionOfTime&>&
+      const std::unordered_map<
+          std::string,
+          std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
           functions_of_time = std::unordered_map<
```

Comment:
is this default required? It seems as though it could potentially lead to confusing behavior. My understanding is that if the derived class is passed as a `CoordinateMap*` and these functions are called on it without the `functions_of_time`, it would then pass the defaulted (empty) map to the derived class implementation, which may or may not try to access the contents depending on how the derived class is supposed to behave.

Reply:
That's right. We want the classes to be able to wrap both time-dependent and time-independent maps. The caller of these maps is responsible for knowing whether the maps are time-dependent. Since this will generically be handled by compute items and nobody should call the maps directly I don't see it causing problems in practice.

Reply:
I see. I suppose if the interface was widely used, I might suggest a solution with more direct overloads function overloads rather than defaults, but it's probably not necessary if the compute items are handling it; regardless I think that sort of change is out of scope for this PR.

Reply:
Sorry, how would an overload by any different from defaults?

Reply:
My thinking was that an overloaded pure virtual would force the derived classes to more explicitly handle the case they're not suited to; the time-dependent ones would have to make choices about what to do when they don't have a `functions_of_time` map (probably just error with a tailored message), so it would be harder to overlook than the current virtuals where I could see a derived class be implemented that just grabs things from the map failing at runtime with more challenging errors. I think It's only a minor improvement, though, so without a better idea, it's probably not worth the time.

Reply:
I think this is a good discussion to have because right now (even in a follow up PR) the interface is still reasonable to change. Let me go over in detail what currently happens:

The `CoordinateMapBase` class only ever has `CoordinateMap<Maps...>` as the derived classes. The `CoordinateMap` class calls the individual coordinate maps, checking whether they require a time argument or not. We could have an `ASSERT` in the functions that calls the coordinate maps that if they are time-dependent that `t` is not the default value (signaling nan) and that the map is not empty. Now, I think checking for a nan with a signalling nan causes an FPE, but that's easily worked around. Hopefully this makes it clear how the maps are working.

In terms of overloading, are you suggesting:
```
operator()(coords);

operator()(coords, time, functions_of_time);
```
? I'm not sure how to make this any different from the current implementation since by the very nature of the base class we erase the knowledge of whether or not the derived maps have any time dependence. That means we can only check at runtime.

Reply:
hm, okay, that brings up other questions.
- It appears that, for instance, `CubicScale` has the same interface that is outlined in `CoordinateMapBase`, but it doesn't inherit, so it must never be used in the generic way that I'm thinking of, but it's unclear to me why. In my suggestion, I was thinking of all specific maps needing to inherit from the `CoordinateMapBase`, but that doesn't seem to be the case.
- It also appears that there are a bunch of overloads (or make_overloads) already in impls for `CoordinateMap...`. If there's already a metafunction for determining time dependence, I think you can use the impl structure to make this a compile-time check with a like the following:
```
TensorType operator()(TensorType coords) noexcept {
    return paren_impl(coords, CoordinateMap_detail::is_map_time_dependent_t<Maps>{});
}
template <typename IsTimeDependent>
TensorType paren_impl(coords, IsTimeDependent /*meta*/) {
    static_assert(not IsTimeDependent::value, "The called map is time dependent, so a function of time must be supplied");
   /// the rest of the function... 
}
```
The reverse could probably be done to prevent a time-independent map from being called with a function of time supplied, but I don't actually know if that version is desirable.

Reply:
- Right, so all coordinate maps will only be used through the `CoordinateMap` class, which inherits off of `CoordinateMapBase`. The motivation was so that rather than having to deal with supporting double/DataVector implementations of virtual functions in all coordinate maps, it's only needed in the "master" versions (`CoordinateMap` and its base). Furthermore, it means fewer places to deal with serializing base/derived classes.
- Yes, so `CoordinateMap` checks each map it composes to see if it should call it with or without the time arguments. We ultimately want to be able to compose both time independent and time-dependent maps together in any order. This means that unless we want to template `Domain` on the individual maps, and thus also the parallel components, we need the maps to be runtime adjustable. What this means is that any such `static_assert` would be triggered all the time since we don't know at compile time whether the maps will be called with or without the arguments. It might be possible to enable such a check without templating everything on the maps, but I would need to think quite a bit more about that to be sure :) The tricky thing is that since the virtual functions can't be templates they will always get instantiated, whether or not they're called, so sticking a static assert in them can easily become `static_assert(false, "Now we never compile");` (I've encountered this and was sad :P )

Reply:
I see. Given that this inheritance is so limited, I'm far less concerned about confusing behavior.

Now, I can see how the set of maps that are called from a particular site might change at runtime; presumably by shuffling around the base class pointers. I also agree that the static assert definitely can't appear in the overridden virtual. But,is it not the case that any case `CoordinateMap_detail::is_map_time_dependent_t<Map>` is `true_type` that the corresponding map functions will runtime error with `unordered_map` access error when called with an empty `unordered_map`? If not, then I've misunderstood the metafunction, but if so it seems like it offers at least a little extra safety to demand at compile time that at least something is passed in for a function of time when it's known that a map that gets called is going to need it.

Reply:
Yes, I think adding `ASSERT`s to check for non-empty functions of time and non-NaN time is a good idea. I've added #1862 and assigned it to me. Once some of the other changes to `CoordinateMap` are in I'll add these checks :)

Reply:
I'm happy with that solution, thanks!

Reply:
You're welcome! Thank you for suggesting it :)